### PR TITLE
[google_maps_flutter]add support for map tapping

### PR DESCRIPTION
**Pull request by RafaO**
Deliver the tap event in the map with the LatLng object where it happened. 

Enables something like this:

          GoogleMap(
            onMapTapped: (LatLng location) {
              showDialog(context: context, builder: (BuildContext context) {
                return AlertDialog(
                  title: new Text("Map tapped!"),
                  content: new Text("You tapped in " + location.latitude.toString() + ", " + location.longitude.toString()),
                  actions: <Widget>[
                    // usually buttons at the bottom of the dialog
                    new FlatButton(
                      child: new Text("Close"),
                      onPressed: () {
                        Navigator.of(context).pop();
                      },
                    ),
                  ],
                );
              });
            },
          )

### CODE REVIEW

**Comment by nandymandy1** [comment]
I m not getting the idea of long tapped event on map to place marker on the tapped point listing or the Coordinates

**Comment by RafaO** [comment]
Hey @nandymandy1 ! This PR is just delivering the tap event. It has nothing to do with long tapped events.

**Comment by nandymandy1** [comment]
how can i get your codes as I m in tight development schedule and how can i get access to your repository???

**Comment by RafaO** [comment]
you can point to my fork like this in your `pubspec.yaml`:
```
  google_maps_flutter:
    git:
      url: git://github.com/RafaO/plugins.git
      ref: integration-branch
      path: packages/google_maps_flutter
```

But do it under your own risk! :)

**Comment by nandymandy1** [comment]
ontap listener ready?????

**Comment by nandymandy1** [comment]
can you show how to use it???

**Comment by RafaO** [comment]
I updated the PR description including an example of usage @nandymandy1 hope it is useful

**Comment by KirolousNashaat** [comment]
can u please update this and re-publish it ?

**Comment by ebeem** [comment]
thanks for your contribution!!
I hope this PR gets merged soon, I believe getting location on tap/long-tap is a very essential feature

I am sorry that you're being asked this frequently, but I will appreciate it if you take some time to update your branch again as some APIs have changed.

also, I tried the snippet you posted, it looks like the listener is on the controller (not the google maps object itself) which seems an excellent decision in my point of view. here is a snippet that will add a marker and print tapped location, please consider updating your snippet so those who want to test your branch do not face any troubles

```Dart
    mapController.onMapTapped.add(
      (LatLng location) {
        print("tapped location ${location.latitude} ${location.longitude}");
        mapController.markers.add(new Marker(
          "location",
          new MarkerOptions(
            icon: BitmapDescriptor.defaultMarker,
            position: location,
          ),
        ));
      },
    );
```

**Comment by RafaO** [comment]
Hi @ebeem Thank you for your message, I also hope this gets merged soon :) 

I am not sure if I understood you completely.

I think this branch is already rebased to the newest changes on master. The controller is no longer used for listeners but the google map object itself. (It used to be the other way around) So I think the snippet is right. Let me know otherwise.

**Comment by ebeem** [comment]
@RafaO what branch are yo actively updating?
it seems like the master is up-to-date, but the integration-branch is not
the master doesn't introduce the new listener onMapTap, while integration-branch does
 but integration branch has some old APIs from old flutter google maps (23 days ago)

```google_maps_flutter```        ```add support for map tapping```        ```23 days ago```

**Comment by RafaO** [comment]
@ebeem the branch I actively update is `map-tap-support` the integration branch is what I use privately in my projects and didn't update it yet. With support for map tapping and this other PR https://github.com/flutter/plugins/pull/981

**Comment by ebeem** [comment]
tested, working great
Thanks

**Comment by iskakaushik** [comment]
@RafaO , I can drive this PR to completion if you do not have the cycles/time to do it. Please allow changes to your branch. Follow the instructions at: https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork

**Comment by RafaO** [comment]
Hi @iskakaushik thanks for your feedback and help. The checkmark for allowing changes from maintainers is already activated (was also this way), I hope you are able to perform changes and of course feel free to do so.

So far I have this:

```
  test('testMapTap', () async {
    LatLng place;
    bool pressed = false;

    final ValueKey<String> mapKey = const ValueKey<String>("map");

    await pumpWidget(Directionality(
        textDirection: TextDirection.ltr,
        child: GoogleMap(
          initialCameraPosition: _kInitialCameraPosition,
          onMapTapped: (LatLng latLng) {
            place = latLng;
            pressed = true;
          },
        ),
        key: mapKey));

    await widgetController.tap(find.byKey(mapKey));

    expect(pressed, true);
  });
```

But is not working, neither with the boolean or with the place being not null.

I am also not sure what is the best way of running these tests because the device stays stuck at the end of tests.

Thank you.

File: packages/google_maps_flutter/lib/src/google_map.dart [hunk 1]
```diff
@@ -32,6 +32,7 @@ class GoogleMap extends StatefulWidget {
     this.onCameraMoveStarted,
     this.onCameraMove,
     this.onCameraIdle,
+    this.onMapTapped,
```
Comment (iskakaushik) [inline packages/google_maps_flutter/lib/src/google_map.dart]: rename this to `onTap` that way it's consistent with `Marker` class.

File: packages/google_maps_flutter/lib/src/google_map.dart [hunk 2]
```diff
@@ -221,6 +225,11 @@ class _GoogleMapState extends State<GoogleMap> {
     final MarkerId markerId = MarkerId(markerIdParam);
     _markers[markerId].infoWindow.onTap();
   }
+
+  void onMapTap(LatLng position) {
```
Comment (iskakaushik) [inline packages/google_maps_flutter/lib/src/google_map.dart]: let's call this `onTap` as well.

File: packages/google_maps_flutter/lib/src/google_map.dart [hunk 3]
```diff
@@ -89,6 +90,9 @@ class GoogleMap extends StatefulWidget {
   /// animations and the user has stopped interacting with the map.
   final VoidCallback onCameraIdle;
 
+  /// Called when map is tapped
```
Comment (iskakaushik) [inline packages/google_maps_flutter/lib/src/google_map.dart]: Consider: `Called every time a [GoogleMap] is tapped.`

**Comment by RafaO** [comment]
Hi @iskakaushik thanks for taking a look and creating the issue. I reworked with your suggestions the PR. Let me know what you think.

**Comment by iskakaushik** [comment]
I think you misread my suggestion as `onMapTap`, I think just `onTap` is better api.

**Comment by iskakaushik** [comment]
Will shortly publish `0.6.0` which should have this. Thanks for this, @RafaO 🥇

**Comment by iskakaushik** [comment]
https://github.com/flutter/plugins/pull/1447

**Comment by kjeremy** [comment]
Finally.

**Pull request merged**